### PR TITLE
Images section

### PR DESCRIPTION
**Pull request by WithoutPants**
Part of #24.

Refactors the scanning logic so that it walks the directory structure instead of globbing the results. Hoping that this will incidentally increase scanning performance further.

Adds customisable file extensions for video, images and gallery files.

Changes the stashes format to support excluding video or image/gallery scanning per stash.

image

Scanning now reads in individual image files.

Adds a new top-level Images section:

image

Wall view uses the flexbin/lightbox from the gallery viewer:

image

The image page is pretty much a direct copy of the scene page:

image

Includes export support, but does not include import support. Import support is waiting on #812.

Schema changes have been made for the new gallery attributes, but no other support is currently implemented.

Other things to do:
* adding images to galleries
* manual creation of galleries
* auto-creation of folder galleries (when selected in the settings)
* scan gallery zip contents to create individual images
* add image tabs to performers/studios

Some of the above can probably be pushed into a separate PR.

I didn't intend to have images have their own page, but for it was quicker and easier to start with the scene page as a base.

### CODE REVIEW

**Comment by bnkai** [comment]
Just from a quick test
Switching from glob star to filepath walk adds a regression for the scanning task.
While the old glob scan traversed through softlinked directories filepath walk by design doesn't.
So users with softlinks in their media paths won't get the same results as before
Related also to issue #706

**Comment by WithoutPants** [comment]
This has turned into a pretty huge change. I'm going to take a break on the main development, to allow some time for testing. It now has manual gallery creation, auto creation/association for folder galleries, and a very clunky add/remove images for galleries. The galleries now have their own page with all of the fields. It's very untested at the moment.

I'm aware of the issue of adding/removing images in galleries not refreshing automatically, but otherwise most of the functionality should be present.

File: graphql/schema/schema.graphql [hunk 1]
```diff
@@ -138,6 +143,28 @@ type Mutation {
   sceneMarkerUpdate(input: SceneMarkerUpdateInput!): SceneMarker
   sceneMarkerDestroy(id: ID!): Boolean!
 
+  imageUpdate(input: ImageUpdateInput!): Image
+  bulkImageUpdate(input: BulkImageUpdateInput!): [Image!]
+  imageDestroy(input: ImageDestroyInput!): Boolean!
+  imagesDestroy(input: ImagesDestroyInput!): Boolean!
+  imagesUpdate(input: [ImageUpdateInput!]!): [Image]
+
+  """Increments the o-counter for a scene. Returns the new value"""
```
Comment (bnkai) [inline graphql/schema/schema.graphql:152]: probably copy paste leftover `image` not `scene`

File: graphql/schema/schema.graphql [hunk 2]
```diff
@@ -138,6 +143,28 @@ type Mutation {
...
+  imageIncrementO(id: ID!): Int!
+  """Decrements the o-counter for a scene. Returns the new value"""
```
Comment (bnkai) [inline graphql/schema/schema.graphql:154]: probably copy paste leftover `image` not `scene`

File: graphql/schema/schema.graphql [hunk 3]
```diff
@@ -138,6 +143,28 @@ type Mutation {
...
+  imageDecrementO(id: ID!): Int!
+  """Resets the o-counter for a scene to 0. Returns the new value"""
```
Comment (bnkai) [inline graphql/schema/schema.graphql:156]: probably copy paste leftover `image` not `scene`

File: pkg/manager/manager_tasks.go [hunk 4]
```diff
@@ -86,6 +78,55 @@ func (t *TaskStatus) updated() {
...
+	// create a control channel through which to signal the counting loop when the timeout is reached
+	chTimeout := time.After(timeout)
+
+	logger.Infof("Counting files to scan...")
+
+	t := 0
+	n := 0
+
+	timeoutErr := errors.New("timed out")
+
```
Comment (bnkai) [inline pkg/manager/manager_tasks.go:93]: `stoppingErr := errors.New("stopping")`
we could use a stopping error also like the scan process for more verbosity

File: pkg/manager/manager_tasks.go [hunk 5]
```diff
@@ -86,6 +78,55 @@ func (t *TaskStatus) updated() {
...
+			//check for timeout
+			select {
+			case <-chTimeout:
+				return timeoutErr
+			default:
+			}
+
+			// check stop
+			if s.Status.stopping {
+				return timeoutErr
```
Comment (bnkai) [inline pkg/manager/manager_tasks.go:111]: `return stoppingErr`

File: pkg/manager/manager_tasks.go [hunk 6]
```diff
@@ -86,6 +78,55 @@ func (t *TaskStatus) updated() {
...
+			}
+
+			return nil
+		})
+
+		if err == timeoutErr {
```
Comment (bnkai) [inline pkg/manager/manager_tasks.go:117]:
```golang
if  ( err == timeoutErr || err == stoppingErr ) {

**Comment by bnkai** [comment]
From a limited testing ( my sample database not the `production` one) everything seems ok, code seems ok also (haven't gone through the mocks or the testing code though) .
A minor regression when you go to a performer page the data panel ends up under the performer image instead of next to it.
The create gallery page also is missing the `add , images` part so you first have to create a gallery and then go to it and add the images (probably what you meant clunky).
I will test against my `production/full` database when possible and leave some more feedback

File: pkg/manager/manager_tasks.go (hunk 4 above)
Comment (WithoutPants) [reply pkg/manager/manager_tasks.go:93]: The behaviour is the same regardless of time out or stopping, and the text itself is not displayed. Logging for stopping is handled in the calling method.

**Comment by cloudstrife11** [comment]
When will this come, Or do u guys need help testing, I have many many images for testing.

**Comment by WithoutPants** [comment]
I still have to do the import code, and I'd like to clean up the gallery image add/remove UI, but yes, it probably needs some testing.

**Comment by bnkai** [comment]
Since transfer.sh crapped out again and there was some interest from the discord channel i made some cross-compilations for anyone that wants to give it a try ( with caution  ofc :-) )
md5s
```bash
f7ac59c6e3edc2be052609a5c589689b  stash-linux
55d43450f9e9ce51bd49be1b3378f1bb  stash-linux-arm32v7
e6505711e566fdf178fa4a64ca8b60a0  stash-linux-arm64v8
0e71209bbc861bcb2f26b0b83615a3ae  stash-osx
c706f4dbad978891ce969cebfe3ebf9a  stash-pi
3669e7612f5efdcf4041fb6430e25da0  stash-win.exe
```
download link
https://easyupload.io/m/r7p0uc

**Comment by bnkai** [comment]
Updated builds
md5s
```bash
39ed0b3a3d9f24372c7cf8bbcde601e8  stash-linux
cd62219392a9926a7063c9e92313a8f5  stash-linux-arm32v7
094e65ad1f83967197f4b07814bc688d  stash-linux-arm64v8
013842d08a5ddbcf0f9212abc1a2b920  stash-osx
3376518c352781744e8587da0e322fd1  stash-pi
1fd80870c82d9e46ac4f5ec18a0e6c65  stash-win.exe
```
https://easyupload.io/m/1jxj84

will try to test over the weekend

**Comment by bnkai** [comment]
updated 
the ui formatting that errored travis shouldn't make much difference
md5s
```bash
589135cc1807b904838d626a4f74a016  stash-linux
c9b231e092e5536101a470bca990e74c  stash-linux-arm32v7
538784094a96a8355e6177e8ee01c1fc  stash-linux-arm64v8
c0174abae820ce7926ec5801db18a161  stash-osx
27c3039e1372dbcdbe56f4fd96e2a823  stash-pi
cb75e44c51a81a81ed12a1fa4623aa16  stash-win.exe
```
https://easyupload.io/m/ldikq1

**Comment by bnkai** [comment]
updated
```bash
2746bddc5ac6c773675eced129eeb025  stash-linux
4fadc5b321194cd8e1a6c3eb8bbf3382  stash-linux-arm32v7
bc6243a97af4ff7f320c2035cbc57ccf  stash-linux-arm64v8
5a245ccddcab92ba6f4257d0bffea876  stash-osx
dad3ddce5c304aaa38c227150eb2a7bd  stash-pi
a313073f21527edaab874aecd499270a  stash-win.exe
```
https://easyupload.io/m/zevbv4

**Pull request merged**